### PR TITLE
Support building Hera VM

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -5,3 +5,7 @@
 [submodule "test/tests"]
 	path = test/jsontests
 	url = https://github.com/ethereum/tests.git
+[submodule "hera"]
+	path = hera
+	url = https://github.com/ewasm/hera
```
Comment (chfast) [inline .gitmodules]: CI cannot find the commit 60963e52d20247376b95e444982453f95a5b300a.
Also, don't use `branch`. It is just a hint for users, and will probably change to master soon.

File: eth/main.cpp [hunk 2]
```diff
@@ -350,8 +350,8 @@ int main(int argc, char** argv)
 	po::options_description generalOptions("General Options", c_lineWidth);
 	generalOptions.add_options()
 		("db-path,d", po::value<string>()->value_name("<path>"), ("Load database from path\n(default: " + getDataDir().string() + ").\n").c_str())
-#if ETH_EVMJIT
-		("vm", "<vm-kind>  Select VM; options are: interpreter, jit or smart (default: interpreter)")
+#if defined(ETH_EVMJIT) || defined(ETH_HERA)
```
Comment (chfast) [inline eth/main.cpp]: I prefer to write it without `defined`. It's still correct and shorter.

File: eth/main.cpp [hunk 3]
```diff
@@ -382,13 +382,13 @@ int main(int argc, char** argv)
...
 
-#if ETH_EVMJIT
+#if defined(ETH_EVMJIT) || defined(ETH_HERA)
 	if (vm.count("vm"))
 	{
 		string vmKind = vm["vm"].as<string>();
 		if (vmKind == "interpreter")
 			VMFactory::setKind(VMKind::Interpreter);
-		else if (vmKind == "jit")
+		else if (vmKind == "jit" || vmKind == "hera")
```
Comment (chfast) [inline eth/main.cpp]: Because the JitVM is not modified for Hera, it will not work at the moment, right?

Comment (axic) [reply eth/main.cpp]: Updated.

File: eth/main.cpp (hunk 2 above)
Comment (axic) [reply eth/main.cpp]: Updated.

File: .gitmodules (hunk 1 above)
Comment (axic) [reply .gitmodules]: Will check out the correct hash once we merge cmake into hera.

File: eth/main.cpp [hunk 4]
```diff
@@ -382,13 +382,13 @@ int main(int argc, char** argv)
 			return -1;
 		}
 
-#if ETH_EVMJIT
+#if ETH_EVMJIT) || ETH_HERA
```
Comment (axic) [inline eth/main.cpp]: Typo, closing )

**Comment by axic** [comment]
@chfast cleaned it up a bit, need to merge proper cmake into hera, then update the reference here. Also probably the cmake changes in this PR need to be reviewed.

File: libevm/VMFactory.cpp [hunk 5]
```diff
@@ -45,7 +45,10 @@ std::unique_ptr<VMFace> VMFactory::create()
 
 std::unique_ptr<VMFace> VMFactory::create(VMKind _kind)
 {
-#if ETH_EVMJIT
+#if ETH_HERA
```
Comment (chfast) [inline libevm/VMFactory.cpp]: This seems to be wrong. It should be `#if ETH_EVMJIT || ETH_HERA`.

File: .gitmodules (hunk 1 above)
Comment (chfast) [reply .gitmodules]: Oh, it works now :)

File: libevm/JitVM.cpp [hunk 6]
```diff
@@ -17,7 +17,11 @@ class EVM
 {
 public:
 	EVM():
+#if EVM_HERA
```
Comment (axic) [inline libevm/JitVM.cpp]: Should be `ETH_HERA`.

File: libevm/CMakeLists.txt [hunk 7]
```diff
@@ -28,3 +28,9 @@ if (EVMJIT)
 	target_link_libraries(evm PRIVATE evmjit)
 	target_compile_definitions(evm PRIVATE ETH_EVMJIT)
 endif()
+
+if (HERA)
+	target_include_directories(evm PRIVATE ../hera)
+	target_link_libraries(evm ${Eth_HERA_LIBRARIES})
+    target_compile_definitions(evm PRIVATE ETH_HERA)
```
Comment (axic) [inline libevm/CMakeLists.txt]: Identation is off?

File: CMakeLists.txt [hunk 8]
```diff
@@ -149,6 +149,10 @@ if (EVMJIT)
     add_subdirectory(evmjit)
 endif()
 
+if (HERA)
+    add_subdirectory(hera)
```
Comment (jwasinger) [inline CMakeLists.txt]: Think we need to add a line here to make it build properly:
```
    set(hera_include_dir ${CMAKE_SOURCE_DIR})
```

This overrides the variable set by hera.

**Comment by axic** [comment]
Still seems to be missing `hera.h`

**Comment by chfast** [comment]
Fixed.

**Comment by axic** [comment]
Appveyor fails:
```
C:\projects\cpp-ethereum\libevm\VMFactory.cpp(136): error C2220: warning treated as error - no 'object' file generated [C:\projects\cpp-ethereum\build\libevm\evm.vcxproj]
C:\projects\cpp-ethereum\libevm\VMFactory.cpp(136): warning C4065: switch statement contains 'default' but no 'case' labels [C:\projects\cpp-ethereum\build\libevm\evm.vcxproj]
```

File: libevm/VMFactory.cpp [hunk 9]
```diff
@@ -19,8 +19,8 @@
 #include "EVMC.h"
 #include "SmartVM.h"
 #include "VM.h"
-#include <libdevcore/Assertions.h>
 #include <evmjit.h>
+#include <hera.h>
```
Comment (winsvega) [inline libevm/VMFactory.cpp]: what if (HERA) flag not defined?

Comment (chfast) [reply libevm/VMFactory.cpp]: It's always defined.

Comment (chfast) [reply libevm/VMFactory.cpp]: The header is always available, because it's in the submodule.

File: libevm/CMakeLists.txt [hunk 10]
```diff
@@ -23,8 +23,14 @@ add_library(evm ${SOURCES} ${HEADERS})
...
+target_include_directories(evm PUBLIC ${CMAKE_SOURCE_DIR}/hera/src)
 
 if (EVMJIT)
 	target_link_libraries(evm PRIVATE evmjit)
 	target_compile_definitions(evm PRIVATE ETH_EVMJIT)
 endif()
+
+if (HERA)
+	target_link_libraries(evm PRIVATE hera)
+    target_compile_definitions(evm PRIVATE ETH_HERA)
```
Comment (axic) [inline libevm/CMakeLists.txt]: Indentation is off?

**Comment by axic** [comment]
@chfast looks good otherwise, fine to merge from me

**Pull request merged**